pages/signin.tsx

```diff
@@ -0,0 +1,156 @@
+import { useState, useEffect } from "react";
+import { useRouter } from "next/router"
+import { Button, CircularProgress, Grid, Paper, Typography, Link, Box, Alert, AlertTitle } from "@mui/material";
+import styles from "../styles/signin.module.css";
+import { GitHubIcon } from "../components/GitHubIcon";
+import * as React from 'react';
+import LoadingButton from '@mui/lab/LoadingButton';
+import TrafficIcon from '@mui/icons-material/Traffic';
+import config from "../config.json"
+
+
+const SCOPE = "repo"
+
+// get code
+function redirect() {
+	const uuid = self.crypto.randomUUID();
+	window.location.href = `https://github.com/login/oauth/authorize?client_id=${config.clientID}&redirect_uri=${config.redirectURL}&scope=${SCOPE}&state=${uuid}`;
+}
+
+type SignInStatus = "calculating" | "not-signed-in" | "exchanging-code-for-token" | "signed-in" | "error-while-signing-in"
+
+
+export default function SignIn() {
+
+	const router = useRouter()
+
+	const [token, setToken] = useState<string>('');
+
+	const [signInStatus, setSignInStatus] = useState<SignInStatus>("calculating")
+
+	const [errorMessage, setErrorMessage] = useState<string>()
+
+	useEffect(() => {
+		setErrorMessage(undefined as any)
+
+		if (router == null) {
+			return
+		}
+		const code = router.query.code?.toString()
+		const error = router.query.error?.toString()
+
+		if (error != null) {
+			setSignInStatus('error-while-signing-in')
+			error === "access_denied" && setErrorMessage("You have denied access to your GitHub account. Please try again.")
+		} else if (code != null) {
+			setSignInStatus('exchanging-code-for-token')
+			console.log(`code ${code}`);
+
+			//This immediately called function must be defined because useEffect handlers cannot be async
+			const isSignedIn = async () => {
+				const response = await fetch(`/api/authenticate/${encodeURIComponent(code.toString())}/`);
+				if (response.status == 200) {
+					return true
+				}
+
+				throw await response.json()
+			};
+			isSignedIn()
+				.then(async () => {
+					setSignInStatus('signed-in')
+				})
+				.catch((error) => {
+
+					console.log("error while exchanging token", error)
+					setSignInStatus('error-while-signing-in')
+					error?.message && setErrorMessage(error.message)
+
+				});
+
+		} else {
+			setSignInStatus('not-signed-in')
+		}
+
+	}, [router]);
+
+	/**
+	 * Called when the user clicks the Try Again button after an error occurred
+	 */
+	function handleTryAgain() {
+		router.push("/signin")
```

Comment:
I'm not sure if this is the intended functionality or because I'm running locally but when I first go to localhost I get this error:
![image](https://user-images.githubusercontent.com/81656054/189239417-2f7f14bc-e366-4339-b585-810e968db745.png)
But then when I click the try again it gives me the option to sign in with github:
![image](https://user-images.githubusercontent.com/81656054/189239493-3ddd1320-3953-4424-bf67-ea2133f77e96.png)
Should it potentially be giving me the option to sign in with github from the start?

Reply:
I think I know why this happens, could you please paste the URL that appears in the browser when you see the error message?

Reply:
The url is `http://localhost:3000/signin?error=Login%20required`
And in my terminal I have these: 
<img width="570" alt="image" src="https://user-images.githubusercontent.com/81656054/189279571-ea2177ec-26c3-4b79-97e6-8086ac85073d.png">
Is it because I don't have CLIENT_Id and CLIENT_SECRET set in my .env?

Reply:
Thanks. Nope, that's just a bug I introduced :)
The dashboard redirects to the signin page if the user is not signed in, adding the `?error=Login%20required` message in the URL. But the signin page displays the error message when the `?error` field is set. 
I'll just remove the `?error` field in the redirect url.